src/com/amazon/corretto/crypto/provider/EvpHmac.java

```diff
@@ -0,0 +1,385 @@
+
+    /**
+     * Returns the size of the array needed to hold the entire HMAC context.
+     */
+    private static native int getContextSize();
+
+    /**
+     * Calls {@code HMAC_Update} with {@code input}, possibly calling {@code HMAC_Init_ex} first
+     * (if {@code evpMd} is any value except {@link #DO_NOT_INIT}).
+     * This method should only be used via {@link #synchronizedUpdateCtxArray(byte[], byte[], long, byte[], int, int)}.
+     *
+     * @param ctx opaque array containing native context
+     */
+    private static native void updateCtxArray(byte[] ctx, byte[] key, long evpMd, byte[] input, int offset, int length);
+    /** @see {@link #updateCtxArray(byte[], byte[], long, byte[], int, int)} */
+    private static void synchronizedUpdateCtxArray(byte[] ctx, byte[] key, long evpMd, byte[] input, int offset, int length) {
+        synchronized(ctx) {
+            updateCtxArray(ctx, key, evpMd, input, offset, length);
+        }
+    }
+
+    /**
+     * Calls {@code HMAC_Final}, and places the result in {@code result}.
+     * This method should only be called via {@link #synchronizedDoFinal(byte[], byte[])}
+     *
+     * @param ctx  opaque array containing native context
+     * @param result
+     */
+    private static native void doFinal(byte[] ctx, byte[] result);
+    /** @see {@link #doFinal(byte[], byte[])} */
+    private static void synchronizedDoFinal(byte[] ctx, byte[] result) {
+        synchronized(ctx) {
+            doFinal(ctx, result);
+        }
+    }
+
+    /**
+     * Calls {@code HMAC_Init_ex}, {@code HMAC_Update}, and {@code HMAC_Final} with
+     * {@code input}.
+     * This method should only be used via
+     * {@link #synchronizedFastHmac(byte[], byte[], long, byte[], int, int, byte[])}.
+     *
+     * @param ctx opaque array containing native context
+     */
+    private static native void fastHmac(
+        byte[] ctx,
+        byte[] key,
+        long evpMd,
+        byte[] input, int offset, int length,
+        byte[] result);
+    /** @see {@link #fastHmac(byte[], byte[], long, byte[], int, int, byte[])} */
+    private static void synchronizedFastHmac(
+            byte[] ctx,
+            byte[] key,
+            long evpMd,
+            byte[] input, int offset, int length,
+            byte[] result) {
+        synchronized(ctx) {
+            fastHmac(ctx, key, evpMd, input, offset, length, result);
+        }
+    }
+
+    private static final int CONTEXT_SIZE = getContextSize();
+
+    // These must be explicitly cloned
+    private HmacState state;
+    private InputBuffer<byte[], Void> buffer;
+
+    EvpHmac(long evpMd, int digestLength) {
+        if (evpMd == DO_NOT_INIT || evpMd == DO_NOT_REKEY) {
+            throw new AssertionError("Unexpected value for evpMd conflicting with reserved negative value: " + evpMd);
+        }
+        this.state = new HmacState(evpMd, digestLength);
+        this.buffer =new InputBuffer<byte[], Void>(1024);
```

Comment:
1024 is arbitrary.

Every time we call from Java into C++ we pay quite a bit of JNI overhead. Thus we want to batch up lots of small updates into fewer larger ones. This is the target batch size. At some point it would be worth doing more detailed analysis to tune it *or* possiblity make it configurable. This value works reasonably well though (from what we've seen experimentally).

Reply:
ok. so this buffer is holding "message" data?

Reply:
Correct. It is holding "message" data.